### PR TITLE
feat(compiler): add break statement support

### PR DESCRIPTION
**Pull request by flyingImer**
Implements Break statement in https://github.com/winglang/wing/issues/114

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingc/src/parser.rs [hunk 1]
```diff
@@ -252,6 +253,17 @@ impl Parser<'_> {
 		})
 	}
 
+	fn build_break_statement(&self, statement_node: &Node) -> DiagnosticResult<StmtKind> {
+		let mut cursor = statement_node.walk();
+		while let Some(parent) = cursor.node().parent() {
```
Comment (flyingImer) [inline libs/wingc/src/parser.rs:258]: for some reason, `cursor.goto_parent()` is always `false`, walking around with this ugly approach. LMK what did I miss

Comment (flyingImer) [reply libs/wingc/src/parser.rs:258]: After playing around with this `goto_parent` API, it works only for the cursor previously went through, e.g., in this use case, it returns `true` IIF we do `cursor.goto_first_child()` first. In other word, it considers the `walk()` TreeNode as the root of the tree. However, I cannot find any doc about this behavior, LMK

**Comment by staycoolcall911** [comment]
Looks great @flyingImer  - I asked @yoav-steinberg to take a look.

File: examples/tests/valid/while.w [hunk 2]
```diff
@@ -6,3 +6,15 @@ let y = 123;
...
 }
+
+let var z = 0;
+while true {
+  z = z + 1;
+  if (z > 2) {
+    break;
+  }
+}
+
```
Comment (eladb) [inline examples/tests/valid/while.w:17]:
```suggestion

assert(z == 3);
```

File: libs/wingc/src/type_check.rs [hunk 3]
```diff
@@ -1542,6 +1542,14 @@ impl<'a> TypeChecker<'a> {
 
 				self.inner_scopes.push(statements);
 			}
+			StmtKind::Break { is_in_loop } => {
+				if !*is_in_loop {
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:1546]: I think this is redundant if we manage to handle the error reporting in _parser.rs_.

Comment (yoav-steinberg) [reply libs/wingc/src/type_check.rs:1546]: We should probably also add an issue for reporting dead code:
```
while true {
  break;
  break; // Should fail on "unreachable code" (out of the scope of this PR)
}
```

**Comment by flyingImer** [comment]
> This, or even if goto_parent would have worked as expected isn't that efficient: each time we encounter a break we'll need to run up the tree. Then if we encounter another break we'll need to do it again. Ideally we'll just have a in_loop: bool in our state (self).
Can you try an approach where you add some mutable state (in_loop) to the Parser class and set it to true when entering a while or for scope. To avoid the complexity of making self mutable lets just make ti a RefCell for now.

This makes a lot of sense, let me take a look

File: libs/wingc/src/type_check.rs (hunk 3 above)
Comment (flyingImer) [reply libs/wingc/src/type_check.rs:1546]: good callout on the dead code! wonder should it fail or just a warning? (I can create a follow-up issue on it) I'm inclined to the latter one - dead code does not hurt anything but it's "dead"

File: libs/wingc/src/parser.rs (hunk 1 above)
Comment (flyingImer) [reply libs/wingc/src/parser.rs:258]: This makes a lot of sense, let me take a look

File: libs/wingc/src/parser.rs [hunk 4]
```diff
@@ -252,6 +253,17 @@ impl Parser<'_> {
...
+			if parent.kind() == "while_statement" || parent.kind() == "for_in_loop" {
+				return Ok(StmtKind::Break { is_in_loop: true });
+			}
+			cursor.reset(parent);
+		}
+		return Ok(StmtKind::Break { is_in_loop: false });
```
[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply libs/wingc/src/parser.rs]: Currently the rule of thumb is to report stuff early on if possible. The actual grammar parsing is done by tree-sitter anyway. _parser.rs_ converts tree-sitter's AST to a strongly typed AST and when doing so does some error checking that the grammar couldn't do on itself (like enum variant uniqueness). Anything we can do without symbol environments and typying I think we should do here.

File: libs/wingc/src/type_check.rs (hunk 3 above)
Comment (yoav-steinberg) [reply libs/wingc/src/type_check.rs]: +1 for an issue on dead-code detection.
This is probably only one of a million other cases we can benefit from this.
We currently don't have warnings in wing, we're trying to see how long this approach will hold water..

Comment (flyingImer) [reply libs/wingc/src/type_check.rs]: dead-code detection tracking: https://github.com/winglang/wing/issues/1767

File: libs/wingc/src/ast.rs [hunk 5]
```diff
@@ -252,9 +252,7 @@ pub enum StmtKind {
 		elif_statements: Vec<ElifBlock>,
 		else_statements: Option<Scope>,
 	},
-	Break {
-		is_in_loop: bool,
-	},
+	Break(),
```
Comment (yoav-steinberg) [inline libs/wingc/src/ast.rs:255]:
```suggestion
	Break,
```

**Review by yoav-steinberg — APPROVED** [review]
excellent!

**Pull request merged**